File: benchmark/groundtruth/tstruth/refs.mjs
// Gabarito TypeScript: definições e referências pelo type checker oficial.
// Uso: node refs.mjs <raiz do repositório> <diretório de saída>
// Grava defs.jsonl e refs.jsonl no mesmo formato do gotruth.
import fs from 'node:fs';
import path from 'node:path';
import ts from 'typescript';

const [rootArg, outArg] = process.argv.slice(2);
if (!rootArg || !outArg) {
  console.error('usage: node refs.mjs <repo root> <out dir>');
  process.exit(2);
}
const root = path.resolve(rootArg);
const out = path.resolve(outArg);
const skipDirs = new Set(['node_modules', '.git', 'dist', 'build', 'coverage', '.next', 'out', '.mira', '.serena', '.turbo']);

function sourceFiles(dir, files = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (!skipDirs.has(entry.name)) sourceFiles(full, files);
    } else if (/\.(ts|tsx|mts|cts)$/.test(entry.name) && !entry.name.endsWith('.d.ts')) {
      files.push(full);
    }
  }
  return files;
}

// Opções do tsconfig da raiz (paths, jsx, baseUrl), sem nada que restrinja a
// lista de arquivos: o programa inclui todo o código do repositório.
function compilerOptions() {
  const defaults = {
    jsx: ts.JsxEmit.Preserve, target: ts.ScriptTarget.ESNext, module: ts.ModuleKind.ESNext,
    moduleResolution: ts.ModuleResolutionKind.Bundler, esModuleInterop: true, resolveJsonModule: true,
  };
  const configPath = path.join(root, 'tsconfig.json');
  let options = defaults;
  if (fs.existsSync(configPath)) {
    const { config } = ts.readConfigFile(configPath, ts.sys.readFile);
    options = { ...defaults, ...ts.parseJsonConfigFileContent(config ?? {}, ts.sys, root).options };
  }
  return { ...options, noEmit: true, skipLibCheck: true, allowJs: false, composite: false, incremental: false, declaration: false, rootDir: undefined };
}

const rel = (file) => path.relative(root, file).split(path.sep).join('/');
const inRepo = (file) => file.startsWith(root + path.sep) && !file.split(path.sep).includes('node_modules');

function isTopLevelVariable(decl) {
  const statement = decl.parent?.parent;
  return Boolean(statement && ts.isVariableStatement(statement) && ts.isSourceFile(statement.parent));
}

// kindOf devolve o tipo das declarações que as ferramentas tratam como
// símbolo; locais, parâmetros e chaves de objeto literal ficam de fora.
function kindOf(decl) {
  switch (decl.kind) {
    case ts.SyntaxKind.FunctionDeclaration: return 'function';
    case ts.SyntaxKind.ClassDeclaration: return 'class';
    case ts.SyntaxKind.InterfaceDeclaration: return 'interface';
    case ts.SyntaxKind.TypeAliasDeclaration: return 'type';
    case ts.SyntaxKind.EnumDeclaration: return 'enum';
    case ts.SyntaxKind.EnumMember: return 'enum_member';
    case ts.SyntaxKind.MethodDeclaration:
    case ts.SyntaxKind.MethodSignature: return 'method';
    case ts.SyntaxKind.PropertyDeclaration:
    case ts.SyntaxKind.PropertySignature: return 'property';
    case ts.SyntaxKind.VariableDeclaration: return isTopLevelVariable(decl) ? 'variable' : '';
    default: return '';
  }
}

function containerOf(decl) {
  const parent = decl.parent;
  if (parent && (ts.isClassLike(parent) || ts.isInterfaceDeclaration(parent) || ts.isEnumDeclaration(parent)) && parent.name) {
    return parent.name.text;
  }
  return undefined;
}

const program = ts.createProgram(sourceFiles(root), compilerOptions());
const checker = program.getTypeChecker();
const defs = new Map();
const refs = new Set();

function position(sf, node) {
  const { line, character } = sf.getLineAndCharacterOfPosition(node.getStart(sf));
  return { file: rel(sf.fileName), line: line + 1, col: character + 1 };
}

function definitionOf(symbol) {
  for (const decl of symbol.declarations ?? []) {
    const kind = kindOf(decl);
    const name = ts.getNameOfDeclaration(decl);
    if (!kind || !name || !inRepo(decl.getSourceFile().fileName)) continue;
    const pos = position(decl.getSourceFile(), name);
    const id = `${pos.file}:${pos.line}:${pos.col}`;
    if (!defs.has(id)) {
      defs.set(id, { id, name: name.getText(decl.getSourceFile()), kind, container: containerOf(decl), ...pos });
    }
    return { id, declarations: symbol.declarations };
  }
  return undefined;
}

function visit(sf, node) {
  if (ts.isIdentifier(node)) {
    let symbol = checker.getSymbolAtLocation(node);
    if (symbol && ts.isShorthandPropertyAssignment(node.parent) && node.parent.name === node) {
      symbol = checker.getShorthandAssignmentValueSymbol(node.parent) ?? symbol;
    }
    if (symbol && symbol.flags & ts.SymbolFlags.Alias) {
      symbol = checker.getAliasedSymbol(symbol);
    }
    const def = symbol && definitionOf(symbol);
    if (def && !def.declarations.some((d) => ts.getNameOfDeclaration(d) === node)) {
      const pos = position(sf, node);
      refs.add(JSON.stringify({ def: def.id, file: pos.file, line: pos.line, col: pos.col }));
    }
  }
  ts.forEachChild(node, (child) => visit(sf, child));
}

for (const sf of program.getSourceFiles()) {
  if (inRepo(sf.fileName)) visit(sf, sf);
}

fs.mkdirSync(out, { recursive: true });
const sortedDefs = [...defs.values()].sort((a, b) => a.id.localeCompare(b.id));
fs.writeFileSync(path.join(out, 'defs.jsonl'), sortedDefs.map((d) => JSON.stringify(d)).join('\n') + '\n');
fs.writeFileSync(path.join(out, 'refs.jsonl'), [...refs].sort().join('\n') + '\n');
console.log(`tstruth: ${sortedDefs.length} definitions, ${refs.size} references`);

File: benchmark/groundtruth/pytruth/refs.py
"""Gabarito Python: definições e referências pelo jedi.

Uso: python refs.py <raiz do repositório> <diretório de saída>
Grava defs.jsonl e refs.jsonl no mesmo formato do gotruth.
"""
import json
import os
import sys

import jedi

SKIP = {".git", "node_modules", ".venv", "venv", "build", "dist", ".mira", ".serena", "__pycache__", ".tox"}


def py_files(root):
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = sorted(d for d in dirnames if d not in SKIP and not d.startswith("."))
        for name in sorted(filenames):
            if name.endswith(".py"):
                yield os.path.join(dirpath, name)


def kind_of(name):
    """Tipo das declarações que as ferramentas tratam como símbolo."""
    parent = name.parent()
    if name.type == "class":
        return "class"
    if name.type == "function":
        return "method" if parent and parent.type == "class" else "function"
    if name.type == "statement" and parent:
        if parent.type == "module":
            return "variable"
        if parent.type == "class":
            return "attribute"
        if parent.type == "function" and f"self.{name.name}" in (name.get_line_code() or ""):
            return "attribute"
    return ""


def container_of(name):
    parent = name.parent()
    if parent and parent.type == "class":
        return parent.name
    if parent and parent.type == "function":
        grand = parent.parent()
        if grand and grand.type == "class":
            return grand.name
    return None


def main():
    root = os.path.abspath(sys.argv[1])
    out = os.path.abspath(sys.argv[2])
    project = jedi.Project(root)
    defs, refs = {}, set()

    def rel(path):
        return os.path.relpath(str(path), root).replace(os.sep, "/")

    def in_repo(path):
        return path is not None and str(path).startswith(root + os.sep)

    def register(name):
        kind = kind_of(name)
        if not kind or not in_repo(name.module_path):
            return None
        ident = f"{rel(name.module_path)}:{name.line}:{name.column + 1}"
        if ident not in defs:
            defs[ident] = {"id": ident, "name": name.name, "kind": kind, "container": container_of(name),
                           "file": rel(name.module_path), "line": name.line, "col": name.column + 1}
        return ident

    files = list(py_files(root))
    for index, path in enumerate(files, 1):
        script = jedi.Script(path=path, project=project)
        for name in script.get_names(all_scopes=True, definitions=True, references=True):
            try:
                targets = name.goto(follow_imports=True, follow_builtin_imports=False)
            except Exception:  # jedi falha em construções raras; a ref só não entra no gabarito
                targets = []
            here = (rel(path), name.line, name.column)
            elsewhere = [t for t in targets if in_repo(t.module_path) and (rel(t.module_path), t.line, t.column) != here]
            if name.is_definition() and not elsewhere:
                register(name)
                continue
            # Um nome de import (`from .app import Flask`) o jedi marca como
            # definição; seguindo o import ele cai na declaração original, e aí
            # é referência, como no gabarito de TypeScript e de Java.
            for target in elsewhere if name.is_definition() else targets:
                ident = register(target)
                if ident and not (rel(path) == defs[ident]["file"] and name.line == defs[ident]["line"]):
                    refs.add((ident, rel(path), name.line, name.column + 1))
        if index % 20 == 0:
            print(f"pytruth: {index}/{len(files)} files", file=sys.stderr, flush=True)

    os.makedirs(out, exist_ok=True)
    with open(os.path.join(out, "defs.jsonl"), "w") as f:
        for ident in sorted(defs):
            f.write(json.dumps(defs[ident]) + "\n")
    with open(os.path.join(out, "refs.jsonl"), "w") as f:
        for ident, file, line, col in sorted(refs):
            f.write(json.dumps({"def": ident, "file": file, "line": line, "col": col}) + "\n")
    print(f"pytruth: {len(defs)} definitions, {len(refs)} references")


if __name__ == "__main__":
    main()
